src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Security.Cryptography;
+
+namespace Internal.Cryptography
+{
+    //
+    // Common infrastructure for crypto symmetric algorithms that use Cng.
+    //
+    internal struct CngSymmetricAlgorithmCore
+    {
+        /// <summary>
+        /// Configures the core to use plaintext keys (to be auto-generated when first needed.)
+        /// </summary>
+        public CngSymmetricAlgorithmCore(string algorithm, ICngSymmetricAlgorithm outer)
+        {
+            _algorithm = algorithm;
+            _outer = outer;
+
+            _keyName = null; // Setting _keyName to null signifies that this object is based on a plaintext key, not a stored CNG key.
+            _provider = null;
+            _optionOptions = CngKeyOpenOptions.None;
+        }
+
+        /// <summary>
+        /// Constructs the core to use a stored CNG key. 
+        /// </summary>
+        public CngSymmetricAlgorithmCore(string algorithm, ICngSymmetricAlgorithm outer, string keyName, CngProvider provider, CngKeyOpenOptions openOptions)
+        {
+            if (keyName == null)
+                throw new ArgumentNullException("keyName");
+            if (provider == null)
+                throw new ArgumentNullException("provider");
+
+            _algorithm = algorithm;
+            _outer = outer;
+
+            _keyName = keyName;
+            _provider = provider;
+            _optionOptions = openOptions;
+
+            using (CngKey cngKey = ProduceCngKey())
+            {
+                CngAlgorithm actualAlgorithm = cngKey.Algorithm;
+                if (algorithm != actualAlgorithm.Algorithm)
+                    throw new CryptographicException(SR.Format(SR.Cryptography_CngKeyWrongAlgorithm, actualAlgorithm.Algorithm, algorithm));
+
+                _outer.BaseKeySize = cngKey.KeySize;
+            }
+        }
+
+        /// <summary>
+        /// Note! This can and likely will throw if the algorithm was given a hardware-based key.
+        /// </summary>
+        public byte[] GetKeyIfExportable()
+        {
+            if (KeyInPlainText)
+            {
+                return _outer.BaseKey;
+            }
+            else
+            {
+                using (CngKey cngKey = ProduceCngKey())
+                {
+                    return cngKey.GetSymmetricKeyDataIfExportable(_algorithm);
+                }
+            }
+        }
+
+        public void SetKey(byte[] key)
+        {
+            _outer.BaseKey = key;
+            _keyName = null; // Setting _keyName to null signifies that this object is now based on a plaintext key, not a stored CNG key.
```

Comment:
I would definitely not vote for (a). Exchanging one software key to another doesn’t propagate the change backward to any other component that happened to be storing the old key. Nor would I expect it to.

(b) Throwing a “Make up your mind” exception if you attempt to set Key or KeySize after constructing the SA with a named key is fine by me
      philosophically, but it isn’t consistent with the overall flavor of SymmetricAlgorithm-based classes so I’d like to get an opinion from the api guys here.





From: Jeremy Barton [mailto:notifications@github.com]
Sent: Monday, October 5, 2015 10:23 AM
To: dotnet/corefx <corefx@noreply.github.com>
Cc: Atsushi Kanamori <atsushik@microsoft.com>
Subject: Re: [corefx] Implement AesCng and TripleDESCng (#3640)


In src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fdotnet%2fcorefx%2fpull%2f3640%23discussion_r41171181&data=01%7c01%7catsushik%40microsoft.com%7c73886730f78044d4f4d908d2cda99f6a%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=V6SZSocWmCPx3Xe91J5iFUURlBQ48lZQRFV1tqfNVXU%3d>:

> +            {

> +                return _outer.BaseKey;

> +            }

> +            else

> +            {

> +                using (CngKey cngKey = ProduceCngKey())

> +                {

> +                    return cngKey.GetSymmetricKeyDataIfExportable(_algorithm);

> +                }

> +            }

> +        }

> +

> +        public void SetKey(byte[] key)

> +        {

> +            _outer.BaseKey = key;

> +            _keyName = null; // Setting _keyName to null signifies that this object is now based on a plaintext key, not a stored CNG key.

Well, the existing things let you change a software key to another software key. In this case, it feels like it should either
a) Update the key in the named key container
b) throw

But
c) switch from a saved (and maybe hardware) key to an unsaved software key

doesn't seem very good. I'm a fan of b; personally.

—
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fdotnet%2fcorefx%2fpull%2f3640%2ffiles%23r41171181&data=01%7c01%7catsushik%40microsoft.com%7c73886730f78044d4f4d908d2cda99f6a%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=7Kqq2xTokfOPQcpCHw%2bNEOWefDJ4QrXGaUouxIgGsfI%3d>.